scrapy_playwright/_utils.py

```diff
@@ -98,34 +97,66 @@ async def _get_header_value(
 
 if platform.system() == "Windows":
 
-    class _WindowsAdapter:
-        """Utility class to redirect coroutines to an asyncio event loop running
-        in a different thread. This allows to use a ProactorEventLoop, which is
-        supported by Playwright on Windows.
+    class _ThreadedLoopAdapter:
+        """Utility class to start an asyncio event loop in a new thread and redirect coroutines.
+        This allows to run Playwright in a different loop than the Scrapy crawler, allowing to
+        use ProactorEventLoop which is supported by Playwright on Windows.
         """
 
-        loop = None
-        thread = None
+        _loop: asyncio.AbstractEventLoop
+        _thread: threading.Thread
+        _coro_queue: asyncio.Queue = asyncio.Queue()
+        _stop_event: asyncio.Event = asyncio.Event()
 
         @classmethod
-        def get_event_loop(cls) -> asyncio.AbstractEventLoop:
-            if cls.thread is None:
-                if cls.loop is None:
-                    policy = asyncio.WindowsProactorEventLoopPolicy()  # type: ignore
-                    cls.loop = policy.new_event_loop()
-                    asyncio.set_event_loop(cls.loop)
-                if not cls.loop.is_running():
-                    cls.thread = threading.Thread(target=cls.loop.run_forever, daemon=True)
-                    cls.thread.start()
-                    logger.info("Started loop on separate thread: %s", cls.loop)
-            return cls.loop
+        async def _handle_coro(cls, coro, future) -> None:
+            try:
+                future.set_result(await coro)
+            except Exception as exc:
+                future.set_exception(exc)
 
         @classmethod
-        async def get_result(cls, coro) -> concurrent.futures.Future:
-            return asyncio.run_coroutine_threadsafe(coro=coro, loop=cls.get_event_loop()).result()
+        async def _process_queue(cls) -> None:
+            while not cls._stop_event.is_set():
+                coro, future = await cls._coro_queue.get()
+                asyncio.create_task(cls._handle_coro(coro, future))
```

Comment:
It may not apply here but, for the record, when implementing session management for scrapy-zyte-api, I found that I needed to [keep live references to task objects until finished](https://github.com/scrapy-plugins/scrapy-zyte-api/blob/cd0aead88570fee360d265d224ccd70196f28fb1/scrapy_zyte_api/_session.py#L765-L767) to prevent them from being garbage-collected.

Reply:
Interesting, thank you very much for the reference. I haven't seen that in my tests but it might be that my runs were not long enough. I won't include it right away just to keep things simple but I'll keep it in mind in case someone reports it as an issue.